### PR TITLE
Create Cocoapods podspec for gRPC C++ repo

### PR DESCRIPTION
**Pull request by muxi**
This is the initial PR for experimental Cocoapods podspec for gRPC C++ library. It allows users to integrate gRPC C++ library in their iOS application with Cocoapods.

Part of #13582.

Note: The podspec only supports build as static library (i.e. no framework support).

Note: the pod only includes native gRPC-C++ without Protobuf. Protobuf support will be implemented in future PR.

### CODE REVIEW

**Comment by grpc-testing** [comment]
```
****************************************************************

libgrpc.so

     VM SIZE        FILE SIZE
 ++++++++++++++  ++++++++++++++

  [ = ]       0        0  [ = ]


****************************************************************

libgrpc++.so

     VM SIZE        FILE SIZE
 ++++++++++++++  ++++++++++++++

  [ = ]       0        0  [ = ]



```

**Comment by grpc-testing** [comment]
```
[trickle] No significant performance differences
```

File: test/cpp/GRPCCppTests/generic/generic.mm [hunk 1]
```diff
@@ -0,0 +1,244 @@
...
+  builder.RegisterAsyncGenericService(&generic_service_);
+  // Include a second call to RegisterAsyncGenericService to make sure that
+  // we get an error in the log, since it is not allowed to have 2 async
+  // generic services
+  builder.RegisterAsyncGenericService(&generic_service_);
+  srv_cq_ = builder.AddCompletionQueue();
+  server_ = builder.BuildAndStart();
+}
+
+- (void)tearDown {
```
Comment (makdharma) [inline test/cpp/GRPCCppTests/generic/generic.mm]: indentation?

File: test/cpp/GRPCCppTests/test/test.m [hunk 2]
```diff
@@ -0,0 +1,43 @@
+//
```
Comment (makdharma) [inline test/cpp/GRPCCppTests/test/test.m]: remove it?

**Comment by grpc-testing** [comment]
```
[microbenchmarks] No significant performance differences
```

**Comment by grpc-testing** [comment]
```
****************************************************************

libgrpc.so

     VM SIZE        FILE SIZE
 ++++++++++++++  ++++++++++++++

  [ = ]       0        0  [ = ]


****************************************************************

libgrpc++.so

     VM SIZE        FILE SIZE
 ++++++++++++++  ++++++++++++++

  [ = ]       0        0  [ = ]



```

**Comment by grpc-testing** [comment]
```
[trickle] No significant performance differences
```

**Comment by grpc-testing** [comment]
```
Corrupt JSON data (indicates timeout or crash): 
    bm_fullstack_unary_ping_pong.BM_UnaryPingPong_MinInProcessCHTTP2_NoOpMutator_NoOpMutator__0_64.counters.new: 1


[microbenchmarks] No significant performance differences
```

File: examples/cpp/helloworld/cocoapods/Podfile [hunk 3]
```diff
@@ -0,0 +1,10 @@
+source 'https://github.com/CocoaPods/Specs.git'
+platform :ios, '8.0'
+
+source 'https://github.com/muxi/grpc-pod.git'
```
Comment (jcanizales) [inline examples/cpp/helloworld/cocoapods/Podfile]: Personal repo

File: examples/cpp/helloworld/cocoapods/HelloWorldCpp/ViewController.mm [hunk 4]
```diff
@@ -0,0 +1,107 @@
...
+ * limitations under the License.
+ *
+ */
+
+#import "ViewController.h"
+#import <grpc++/grpc++.h>
+#include <grpc++/generic/generic_stub.h>
+#include <grpc++/generic/async_generic_service.h>
+
+void* tag(int i) { return (void*)(intptr_t)i; }
```
Comment (jcanizales) [inline examples/cpp/helloworld/cocoapods/HelloWorldCpp/ViewController.mm]: `static void*`

Comment (jcanizales) [reply examples/cpp/helloworld/cocoapods/HelloWorldCpp/ViewController.mm]: And I think the C++ convention is `Tag`, but I've never looked at the ObjC++ style guide.

File: examples/cpp/helloworld/cocoapods/HelloWorldCpp/ViewController.mm [hunk 5]
```diff
@@ -0,0 +1,107 @@
...
+
+// Serialized Proto bytes of Hello World example
+const uint8_t message[] =
```
Comment (jcanizales) [inline examples/cpp/helloworld/cocoapods/HelloWorldCpp/ViewController.mm]: `kMessage`

File: examples/cpp/helloworld/cocoapods/HelloWorldCpp/ViewController.mm [hunk 6]
```diff
@@ -0,0 +1,107 @@
...
+    {0x0A, 0x0B, 0x4F, 0x62, 0x6A, 0x65, 0x63, 0x74, 0x69, 0x76, 0x65, 0x2D, 0x43};
+
+@interface ViewController ()
+
+@end
+
+@implementation ViewController {
+  grpc::CompletionQueue cli_cq_;
```
Comment (jcanizales) [inline examples/cpp/helloworld/cocoapods/HelloWorldCpp/ViewController.mm]: What's CLI? Command-line interface?

File: examples/cpp/helloworld/cocoapods/HelloWorldCpp/ViewController.mm [hunk 7]
```diff
@@ -0,0 +1,107 @@
...
+          [recvBytes stringByAppendingString:[NSString stringWithFormat:@"%02x ", *p]];
+      p++;
+    }
+  }
+  NSLog(@"Hello World succeeded.\nReceived bytes: %@\n"
+        "Expected bytes: 0a 11 48 65 6c 6c 6f 20 4f 62 6a 65 63 74 69 76 65 2d 43", recvBytes);
+}
+
+
+- (void)didReceiveMemoryWarning {
```
Comment (jcanizales) [inline examples/cpp/helloworld/cocoapods/HelloWorldCpp/ViewController.mm]: You can remove this as it's not doing anything.

File: test/cpp/GRPCCppTests/Podfile [hunk 8]
```diff
@@ -0,0 +1,71 @@
...
+#
+# This podspec searches for the gRPC core library headers under "$(PODS_ROOT)/gRPC-Core", where
+# Cocoapods normally places the downloaded sources. When doing local development of the libraries,
+# though, Cocoapods just takes the sources from whatever directory was specified using `:path`, and
+# doesn't copy them under $(PODS_ROOT). When using static libraries, one can sometimes rely on the
+# symbolic links to the pods headers that Cocoapods creates under "$(PODS_ROOT)/Headers". But those
+# aren't created when using dynamic frameworks. So our solution is to modify the podspec on the fly
+# to point at the local directory where the sources are.
+#
+# TODO(jcanizales): Send a PR to Cocoapods to get rid of this need.
```
Comment (jcanizales) [inline test/cpp/GRPCCppTests/Podfile]: We should link to an issue here.

File: gRPC-C++.podspec [hunk 9]
```diff
@@ -0,0 +1,679 @@
...
+# limitations under the License.
+
+
+Pod::Spec.new do |s|
+  s.name     = 'gRPC-C++'
+  # TODO (mxyan): use version that match gRPC version when pod is stabilized
+  # version = '1.9.0-dev'
+  version = '0.0.1'
+  s.version  = version
+  s.summary  = 'gRPC C++ library'
```
Comment (jcanizales) [inline gRPC-C++.podspec]: C++ runtime?

File: gRPC-C++.podspec [hunk 10]
```diff
@@ -0,0 +1,679 @@
...
+                      'src/core/ext/transport/chttp2/transport/varint.h',
+                      'src/core/ext/transport/chttp2/alpn/alpn.h',
+                      'src/core/ext/filters/http/client/http_client_filter.h',
+                      'src/core/ext/filters/http/message_compress/message_compress_filter.h',
+                      'src/core/ext/filters/http/server/http_server_filter.h'
+  end
+
+  s.prepare_command = <<-END_OF_COMMAND
+    find src/cpp/ -type f -exec sed -E -i '.back' 's;#include "third_party/nanopb/(.*)";#include <nanopb/\\1>;g' {} \\\;
+    find src/cpp/ -name "*.back" -type f -delete
```
Comment (jcanizales) [inline gRPC-C++.podspec]: Can't you tell find not to save a backup?

File: test/cpp/GRPCCppTests/test/server_context_test_spouse_test.mm [hunk 11]
```diff
@@ -0,0 +1,106 @@
...
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ *
+ */
+
+#define TEST(a,b) - (void)test ## b
```
Comment (jcanizales) [inline test/cpp/GRPCCppTests/test/server_context_test_spouse_test.mm]: If `a` is ignored on purpose, clarify why in a comment :)

File: test/cpp/GRPCCppTests/test/server_context_test_spouse_test.mm [hunk 12]
```diff
@@ -0,0 +1,106 @@
...
+  context.AddTrailingMetadata(key1, val1);
+  metadata.insert(std::pair<grpc::string, grpc::string>(key1, val1));
+  ASSERT_EQ(metadata, spouse.GetTrailingMetadata());
+
+  context.AddTrailingMetadata(key2, val2);
+  metadata.insert(std::pair<grpc::string, grpc::string>(key2, val2));
+  ASSERT_EQ(metadata, spouse.GetTrailingMetadata());
+}
+
+- (void)setUp {
```
Comment (jcanizales) [inline test/cpp/GRPCCppTests/test/server_context_test_spouse_test.mm]: No need to override if nothing's done.

File: test/cpp/GRPCCppTests/generic/generic.mm [hunk 13]
```diff
@@ -0,0 +1,244 @@
...
+
+#include "test/core/util/port.h"
+#include "test/core/util/test_config.h"
+
+using std::chrono::system_clock;
+using namespace grpc;
+
+void* tag(int i) { return (void*)(intptr_t)i; }
+
+static grpc_slice merge_slices(grpc_slice* slices, size_t nslices) {
```
Comment (jcanizales) [inline test/cpp/GRPCCppTests/generic/generic.mm]: Move this and the next function to a shared test-util library? We can't be the only ones with a need for it.

File: test/cpp/GRPCCppTests/test/server_context_test_spouse_test.mm [hunk 14]
```diff
@@ -0,0 +1,106 @@
...
+    [super setUp];
```
Comment (jcanizales) [inline test/cpp/GRPCCppTests/test/server_context_test_spouse_test.mm]: Why do you need `grpc_test_init` in `generic.mm` but not here?

File: gRPC-C++.podspec (hunk 9 above)
Comment (muxi) [reply gRPC-C++.podspec]: I think 'library' is fine?

File: gRPC-C++.podspec (hunk 10 above)
Comment (muxi) [reply gRPC-C++.podspec]: -i with nothing works for GNU but not for Mac. No universal solution.

File: test/cpp/GRPCCppTests/Podfile (hunk 8 above)
Comment (muxi) [reply test/cpp/GRPCCppTests/Podfile]: Will keep it the same as the old one for now and remove both together with Cocoapods' fix.

File: test/cpp/GRPCCppTests/generic/generic.mm (hunk 13 above)
Comment (muxi) [reply test/cpp/GRPCCppTests/generic/generic.mm]: I would prefer not to structure test utils until we really need to. Original cpp tests used some other mechanisms and do not need this too.

File: test/cpp/GRPCCppTests/test/server_context_test_spouse_test.mm (hunk 14 above)
Comment (muxi) [reply test/cpp/GRPCCppTests/test/server_context_test_spouse_test.mm]: Not sure why, but it was not in the original test. Probably it does not need to initiate random seed in this case.

**Comment by grpc-testing** [comment]
```
****************************************************************

libgrpc.so

     VM SIZE        FILE SIZE
 ++++++++++++++  ++++++++++++++

  [ = ]       0        0  [ = ]


****************************************************************

libgrpc++.so

     VM SIZE        FILE SIZE
 ++++++++++++++  ++++++++++++++

  [ = ]       0        0  [ = ]



```

**Comment by grpc-testing** [comment]
```
[trickle] No significant performance differences
```

**Comment by grpc-testing** [comment]
```
[microbenchmarks] No significant performance differences
```

**Comment by grpc-testing** [comment]
```
****************************************************************

libgrpc.so

     VM SIZE        FILE SIZE
 ++++++++++++++  ++++++++++++++

  [ = ]       0        0  [ = ]


****************************************************************

libgrpc++.so

     VM SIZE        FILE SIZE
 ++++++++++++++  ++++++++++++++

  [ = ]       0        0  [ = ]



```

**Comment by grpc-testing** [comment]
```
[trickle] No significant performance differences
```

**Comment by grpc-testing** [comment]
```
[microbenchmarks] No significant performance differences
```

**Pull request merged**